### PR TITLE
groups: implemented .subgroup() method

### PR DESCRIPTION
**Pull request by valglad**
This PR implements a method `.subgroup()` for groups (both permutation and `FpGroup`s) which takes the subgroup generators as the argument and returns another group whose `parent` is set to be the original group.

* `PermutationGroup`s: the method simply creates another `PermutationGroup` on the given generators. The use of `.parent` in the `.is_subgroup()` method allows for faster subgroup checking in some cases.

* `FpGroup`s: `reidemeister_presentation()` method returns the generators and relators for a presentation of the subgroup. These are instances of `FreeGroupElement`. To return an `FpGroup` based on them, it is necessary to create a `FreeGroup` on the generators first and make sure that the generators of this free group match up with those used in the relators. For this purpose I made it possible to create a `FreeGroup` by passing instances of `FreeGroupElement`s to it. 

The other changes are relatively minor and not directly relevant to the new method:
* Generators of `FpGroup` are returned by the property `.generators` while a method `.relators()` has to be called for the relators. It feels a bit inconsistent so I made `.relators` a property.

* Rearranged things in the file `fp_groups.py` to have properties all together near the top of the class definition.

* Corrected a couple of typos, added a missing error message and slightly changed the description for the arguments of the `FreeGroup` class.

### CODE REVIEW

File: sympy/combinatorics/free_groups.py [hunk 1]
```diff
@@ -94,14 +94,17 @@ def _parse_symbols(symbols):
...
         return (symbols,)
     elif is_sequence(symbols):
         if all(isinstance(s, string_types) for s in symbols):
             return _symbols(symbols)
-        elif all(isinstance(s, Expr) for s in symbols):
+        elif all(isinstance(s, Expr) for s in symbols) or \
+             all(isinstance(s, FreeGroupElement) for s in symbols):
             return symbols
-    raise ValueError("")
+    raise ValueError("The type of `symbols` must be one of the following: "\
```
Comment (gxyd) [inline sympy/combinatorics/free_groups.py]: I think, having `\` at the end wouldn't be necessary (better not to have them).

File: sympy/combinatorics/free_groups.py [hunk 2]
```diff
@@ -94,14 +94,17 @@ def _parse_symbols(symbols):
         return tuple()
     if isinstance(symbols, string_types):
         return _symbols(symbols, seq=True)
-    elif isinstance(symbols, Expr):
+    elif isinstance(symbols, Expr) or isinstance(symbols, FreeGroupElement):
```
Comment (jksuom) [inline sympy/combinatorics/free_groups.py]: A single `isinstance` should suffice.

File: sympy/combinatorics/fp_groups.py [hunk 3]
```diff
@@ -69,13 +69,59 @@ def __new__(cls, fr_grp, relators):
...
+
+    def subgroup(self, gens):
+        '''
+        Return the subgroup generated by `gens` using the
+        Reidemeister-Schreier algorithm
+
+        '''
+        if not all([isinstance(g, FreeGroupElement) for g in gens]):
+            raise ValueError("Generators must be `FreeGroupElement`s")
+        letters = [g.array_form[0][0] for g in self.generators]
```
Comment (jksuom) [inline sympy/combinatorics/fp_groups.py]: Is this the same as `self.symbols`?

Comment (valglad) [inline sympy/combinatorics/fp_groups.py]: `self` is an `FpGroup` and it doesn't have `.symbols`. Though `self.free_group.symbols` is probably the same.

File: sympy/combinatorics/fp_groups.py [hunk 4]
```diff
@@ -69,13 +69,59 @@ def __new__(cls, fr_grp, relators):
...
+        Return the subgroup generated by `gens` using the
+        Reidemeister-Schreier algorithm
+
+        '''
+        if not all([isinstance(g, FreeGroupElement) for g in gens]):
+            raise ValueError("Generators must be `FreeGroupElement`s")
+        letters = self.free_group.symbols
+        for g in gens:
+            c = [t[0] in letters for t in g.array_form]
+            if not all(c):
```
Comment (jksuom) [inline sympy/combinatorics/fp_groups.py]: Each FreeGroupElement has attribute `group`. Perhaps we should use the condition `g.group == self.free_group`.

Comment (valglad) [inline sympy/combinatorics/fp_groups.py]: That would be good but that also makes me realise that we can't just make a new `FreeGroup` on already existing `FreeGroupElement`s because then the generators end up with a wrong `.group` attribute. Also, the `group` attribute of the generators returned by `reidemester_presentation()` is a free group on more generators than necessary. Hmm, I'll have another look at it. Having `reidemester_presentation()` return generators with the right `group` attribute and matching relators would probably be better. Alternatively, go through the generators and relators changing their `.group` attribute when a new `FreeGroup` is created on them.

File: sympy/combinatorics/perm_groups.py [hunk 5]
```diff
@@ -3018,6 +3027,19 @@ def strong_gens(self):
...
+        """
+           Return the subgroup generated by `gens` which is a list of
+           elements of the group
+        """
+
+        if not all([g in self for g in gens]):
+            raise ValueError("The group doesn't contain the supplied generators")
+
+        G = PermutationGroup(gens)
+        G._parent = self
```
Comment (jksuom) [inline sympy/combinatorics/perm_groups.py]: For what purpose is this defined?

File: sympy/combinatorics/tests/test_fp_groups.py [hunk 6]
```diff
@@ -808,6 +808,9 @@ def test_subgroup_presentations():
     p1 = reidemeister_presentation(f, H)
     assert str(p1) == "((y_1, y_2), (y_1**2, y_2**3, y_2*y_1*y_2*y_1*y_2*y_1))"
 
+    H = f.subgroup(H)
+    assert (H.generators == p1[0]) and (H.relators == p1[1])
```
Comment (jksuom) [inline sympy/combinatorics/tests/test_fp_groups.py]: This could probably have `(H.generators, H.relators) == p1`.

File: sympy/combinatorics/perm_groups.py (hunk 5 above)
Comment (valglad) [inline sympy/combinatorics/perm_groups.py]: I suppose it's not as relevant to permutation groups. For `FpGroup`s it allows to have `is_subgroup` method, and also this is a set-up GAP has (i.e. having `.parent` attribute for subgroups of groups). Theoretically having this might save time when someone tries to call `is_subgroup` on large permutation groups but mainly I was just mimicking what I did for `FpGroup`s here. Should I remove it?

Comment (jksuom) [inline sympy/combinatorics/perm_groups.py]: I think you could do that.

File: sympy/combinatorics/fp_groups.py [hunk 7]
```diff
@@ -76,6 +76,27 @@ def __new__(cls, fr_grp, relators):
     def free_group(self):
         return self._free_group
 
+    @property
+    def relators(self):
+        return tuple(self._relators)
```
Comment (jksuom) [inline sympy/combinatorics/fp_groups.py]: Would there be any reason why we couldn't have `self.relators` instead of `self._relators` as a plain attribute?

File: sympy/combinatorics/fp_groups.py [hunk 8]
```diff
@@ -1712,6 +1726,19 @@ def reidemeister_presentation(fp_grp, H, elm_rounds=2, simp_rounds=2):
     for i in range(20):
         elimination_technique_1(C)
         simplify_presentation(C)
+
+    syms = [g.array_form[0][0] for g in C._schreier_generators]
+    g = free_group(syms)[0]
```
Comment (jksuom) [inline sympy/combinatorics/fp_groups.py]: How is this related to `C._schreier_free_group`?

File: sympy/combinatorics/fp_groups.py (hunk 7 above)
Comment (valglad) [inline sympy/combinatorics/fp_groups.py]: I don't think so. And I don't think there is any reason to have `self._free_group` as well as `self.free_group`. Should I change both of them?

File: sympy/combinatorics/fp_groups.py (hunk 8 above)
Comment (valglad) [inline sympy/combinatorics/fp_groups.py]: It's a free group on generators that have the same symbols as a subset of `C._schreier_free_group`'s generators.

File: sympy/combinatorics/fp_groups.py (hunk 7 above)
Comment (jksuom) [inline sympy/combinatorics/fp_groups.py]: I think you could do that. For `relators` in particular, that seems better than defining a property. But it would probably be preferable to do the changes in a separate PR.
